kiwi/xml_state.py

```diff
@@ -522,33 +522,52 @@ def get_users(self):
         """
         List of configured users.
 
-        Each entry in the list is a tuple with the following information
-
-        * group_name, name of the group
-        * group_id, id of the group
-        * user_sections, list of xml_parse::users instances which belong
-          to the group name and id of this tuple
+        Each entry in the list is a single xml_parse::user instance.
 
         :return: user data
         :rtype: list
         """
-        users_by_group_type = namedtuple(
-            'users_by_group_type', ['group_name', 'group_id', 'user_sections']
-        )
-        users_by_group = []
+        users_list = []
         users_sections = self.get_users_sections()
-        if users_sections:
-            for users in users_sections:
-                user_sections = users.get_user()
-                if user_sections:
-                    users_by_group.append(
-                        users_by_group_type(
-                            group_name=users.get_group(),
-                            group_id=users.get_id(),
-                            user_sections=user_sections
-                        )
-                    )
-        return users_by_group
+        if len(users_sections):
+            for users_section in users_sections:
+                for user in users_section.get_user():
+                    match = [i for i, u in enumerate(users_list) if u.get_name() == user.get_name()]
+                    for idx in match:
+                        if users_list[idx].get_groups() and user.get_groups():
+                            users_list[idx].set_groups(
+                                ','.join([users_list[idx].get_groups(), user.get_groups()])
+                            )
+                        elif not users_list[idx].get_groups() and user.get_groups():
+                            users_list[idx].set_groups(user.get_groups())
+                    if not len(match):
+                        users_list.append(user)
+
+        return users_list
+
+    def get_user_groups(self, user_name):
+        """
+        List of configured users.
+
+        Each entry in the list is the name of a group that the specified
+        user belongs to. The first item in the list is the login or primary
+        group. The list will be empty if no groups are specified in the
+        description file.
+
+        :return: groups data for the given user
+        :rtype: list
+        """
+        groups_list = []
+        for user in self.get_users():
+            if user.get_name() == user_name and user.get_groups():
+                seen = set()
+                groups_list = [
+                    grp for grp in user.get_groups().split(',')
+                    if not (grp in seen or seen.add(grp))
+                ]
+                break
+
+        return groups_list
```

Comment:
The problem here is that if we assume to get a list of unique users from `self.get_users()` then we need to have the groups appended in advance (that the motivation to modify groups attribute inside get_users).

Also using `set(groups_list)` does not keep the order groups, so we have no clear criteria on which group is going to be to login group.

Maybe we can assume self.get_users() to return more than one instance of the same user, that would lead to a  user modification so the user would be configured as it is in the last appearance. That could simplify this code.